lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -940,8 +948,11 @@ class ADContext {
   /// parameters as the function.
   StructDecl *createPrimalValueStruct(const DifferentiationTask *task);
 
-  void insertGradient(const GradientLookupKey &key, SILFunction *gradient) {
+  void insertGradient(const GradientLookupKey &key, SILFunction *gradient,
```

Comment:
You don't need a separate flag to distinguish between canonical and non-canonical. Instead, use `key.second.isMaster()`.